File: 1800-concatenation-of-consecutive-binary-numbers/concatenation-of-consecutive-binary-numbers.py
class Solution:
    def concatenatedBinary(self, n: int) -> int:
        result = 0
        MOD = 10**9 + 7
        bit_length = 0
        
        for i in range(1, n + 1):
            # Check if i is a power of 2 (1, 2, 4, 8, 16...)
            # Every time we hit a power of 2, i needs one more bit to represent it
            if i & (i - 1) == 0:
                bit_length += 1
            
            # 1. Shift current result left by the number of bits in i
            # 2. Use bitwise OR (|) to append i to the end
            # 3. Take modulo to keep the number within 10^9 + 7
            result = ((result << bit_length) | i) % MOD
            
        return result

File: 1800-concatenation-of-consecutive-binary-numbers/concatenation-of-consecutive-binary-numbers.java
class Solution {
    public int concatenatedBinary(int n) {
        long result = 0;
        int MOD = 1_000_000_007;
        int bitLength = 0;

        for (int i = 1; i <= n; i++) {
            // If i is a power of 2, the binary representation 
            // needs one more bit (e.g., 2 is "10", 4 is "100")
            if ((i & (i - 1)) == 0) {
                bitLength++;
            }
            
            // 1. Shift the current result left by the length of i
            // 2. Use bitwise OR to place i in the new space
            // 3. Apply modulo to keep the number manageable
            result = ((result << bitLength) | i) % MOD;
        }

        return (int) result;
    }
}